google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py

```diff
@@ -168,17 +168,30 @@ def ack(self, items: Sequence[requests.AckRequest]) -> None:
 
         # We must potentially split the request into multiple smaller requests
         # to avoid the server-side max request size limit.
-        ack_ids = (item.ack_id for item in items)
+        items_gen = (i for i in items)
+        ack_ids_gen = (item.ack_id for item in items)
         total_chunks = int(math.ceil(len(items) / _ACK_IDS_BATCH_SIZE))
 
         for _ in range(total_chunks):
-            request = gapic_types.StreamingPullRequest(
-                ack_ids=itertools.islice(ack_ids, _ACK_IDS_BATCH_SIZE)
-            )
-            self._manager.send(request)
-
-        # Remove the message from lease management.
-        self.drop(items)
+            future_reqs_dict = {req.ack_id: req for req in itertools.islice(items_gen, _ACK_IDS_BATCH_SIZE) if req.future}
+            requests_completed, requests_to_retry =  self._manager.send_unary_ack(
+                ack_ids=list(itertools.islice(ack_ids_gen, _ACK_IDS_BATCH_SIZE)),
+                future_reqs_dict=future_reqs_dict)
+
+            # Remove the completed messages from lease management.
+            self.drop(requests_completed)
+
+            # retry acks
+            # todo: figure out if retry delays/limits are needed
+            while requests_to_retry:
+                _LOGGER.debug("Retrying " + str(len(requests_to_retry)) + " acks")
+                future_reqs_dict = {req.ack_id: req for req in requests_to_retry if req.future}
+                requests_completed, requests_to_retry = self._manager.send_unary_ack(
+                    ack_ids=[req.ack_id for req in requests_to_retry],
+                    future_reqs_dict=future_reqs_dict)
+                assert len(requests_to_retry) <= _ACK_IDS_BATCH_SIZE, "Too many requests to be retried."
```

Comment:
Is this meant to be used as a runtime check in production, i.e. not just when running tests?

If so, an exception should be explicitly raised if the invariant does not hold. The problem with `assert` statements is that they are removed if Python is run in optimized mode, demo:
```
$ python -c "x = 0; assert x != 0; print(1/x)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AssertionError

$ python -O -c "x = 0; assert x != 0; print(1/x)"  # optimized mode
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ZeroDivisionError: division by zero

```

Reply:
Good to know! I intend it to be closer to invariant documentation. I'm not relying on it to run in prod.